src/v/ssx/fiber_local_storage.h

```diff
@@ -0,0 +1,174 @@
+#include "seastarx.h"
+#include "ssx/this_fiber_id.h"
+#include "utils/intrusive_list_helpers.h"
+#include "vassert.h"
+
+#include <seastar/core/future.hh>
+
+namespace internal {
+
+template<class Tag, class ValueT>
+class fiber_local_impl {
+public:
+    template<class T>
+    requires(
+      !std::is_same_v<
+        T,
+        fiber_local_impl<Tag, ValueT>>) explicit fiber_local_impl(T&& value)
+      : _id(ssx::this_fiber_id())
+      , _value(std::forward<T>(value)) {
+        _fifo.push_front(std::ref(*this));
+    }
+
+    fiber_local_impl()
+      : _id(ssx::this_fiber_id())
+      , _value() {
+        _fifo.push_front(std::ref(*this));
+    }
+
+    fiber_local_impl(const fiber_local_impl&) = delete;
+    fiber_local_impl(fiber_local_impl&&) = delete;
+    fiber_local_impl& operator=(const fiber_local_impl&) = delete;
+    fiber_local_impl& operator=(fiber_local_impl&&) = delete;
+
+    static fiber_local_impl<Tag, ValueT>* get_fiber_local() {
+        auto id = ssx::this_fiber_id();
+        auto it = std::find_if(
+          _fifo.begin(),
+          _fifo.end(),
+          [id](const fiber_local_impl<Tag, ValueT>& it) {
+              if (it._id == id) {
+                  return true;
+              }
+              return false;
+          });
+        return it == _fifo.end() ? nullptr : &(*it);
+    }
+
+    template<class T>
+    void set(T&& val) noexcept {
+        _value = std::forward<T>(val);
+    }
+
+    const ValueT& get() const noexcept { return _value; }
+
+private:
+    uint64_t _id;
+    intrusive_list_hook _hook;
+    ValueT _value;
+
+    using intr_list_t
+      = intrusive_list<fiber_local_impl, &fiber_local_impl::_hook>;
+
+    /// Thread local storage for the fiber
+    static thread_local intr_list_t _fifo;
+};
+
+template<class Tag, class ValueT>
+thread_local typename fiber_local_impl<Tag, ValueT>::intr_list_t
+  fiber_local_impl<Tag, ValueT>::_fifo
+  = intr_list_t();
+
+} // namespace internal
+
+namespace ssx {
+
+/// Fiber local storage instance.
+///
+/// This is a container type that can hold object of any
+/// type which becomes available for any code that runs in
+/// the same fiber.
+/// Any fiber_local object with the same tag type will have the same content.
+/// Even if the fiber_local object is recreated the data that it holds will
+/// still be available.
```

Comment:
What is this line promising about `fiber_local`? Is it the following behavior? 

```
fiber_local<my_tag_type_t, int> fiber(1);
{
    fiber_local<my_tag_type_t, int> fiber(2);
    fiber_local_selector<my_tag_type_t, int> selector;
    assert(selector.get() == 2);
}
fiber_local_selector<my_tag_type_t, int> selector;
assert(selector.get() == 1);
```
As far as I can tell both of these assertions will be true in the current implementation.

Reply:
Noticed local shadowing is a test case. Resolving conversation.